Content.Shared/_EE/Items/OniOnlySystem.cs

```diff
@@ -0,0 +1,52 @@
+using Content.Server.Abilities.Oni;
+using Content.Shared.Ghost;
+using Content.Shared.Hands.EntitySystems;
+using Content.Shared.Popups;
+using Content.Shared.StatusEffect;
+using Content.Shared.Stunnable;
+using Content.Shared.Weapons.Melee.Events;
+using Robust.Shared.Audio;
+using Robust.Shared.Audio.Systems;
+
+
+namespace Content.Shared._EE.Item;
+
+public sealed class OniOnlySystem : EntitySystem
+{
+    [Dependency] private readonly SharedStunSystem _stun = default!;
+    [Dependency] private readonly SharedPopupSystem _popupSystem = default!;
+    [Dependency] private readonly SharedHandsSystem _hands = default!;
+    [Dependency] private readonly StatusEffectsSystem _statusEffects = default!;
+    [Dependency] private readonly SharedAudioSystem _audioSystem = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+        SubscribeLocalEvent<OniOnlyComponent, AttemptMeleeEvent>(OnMeleeAttempt);
+    }
+
+    private void OnMeleeAttempt(EntityUid uid, OniOnlyComponent component, ref AttemptMeleeEvent args)
+    {
+        bool CanUse(EntityUid? uid) => HasComp<OniComponent>(uid) || HasComp<GhostComponent>(uid);
```

Comment:
```suggestion
    private bool bool CanUse(EntityUid uid) => HasComp<OniComponent>(uid) || HasComp<GhostComponent>(uid);

    private void OnMeleeAttempt(EntityUid uid, OniOnlyComponent component, ref AttemptMeleeEvent args)
    {
```
This really should actually be checking against a ComponentRegistry stored in a datafield on the component. Check up on how TraitSystem.Functions makes use of ComponentRegistry datafields if you need a reference on how it works. The gist of it in this context is that you would Foreach the datafield, and return true the first time HasComp declares that the entity has the component being checked. Otherwise return false after the Foreach is done.